### PR TITLE
Adding initial SVE support to TVM

### PR DESCRIPTION
**Pull request by MeeraN7**
Prototype for the addition of Arm Architecture's Scalable Vector Extension (SVE) in TVM containing initial VLA and predication implementation, based on earlier work by Giuseppe Rossini. 

The RFC can be found [here](https://github.com/apache/tvm-rfcs/pull/18)

### CODE REVIEW

**Comment by MeeraN7** [comment]
@sjoerdmeijer @giuseros @mbaret

**Comment by mbaret** [comment]
@tqchen @jcf94 @junrushao1994 you may also be interested

File: src/target/llvm/codegen_aarch64.cc [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+      break;
+    case 4:  // float
+      cnt_id = llvm::Function::lookupIntrinsicID("llvm.aarch64.sve.cntw");
+      break;
+    default:  // double
+      cnt_id = llvm::Function::lookupIntrinsicID("llvm.aarch64.sve.cntd");
+  }
+
+  // All pattern
+  int all_pattern = 31;
```
Comment (huajsj) [inline src/target/llvm/codegen_aarch64.cc:95]: use macro?

File: include/tvm/runtime/data_type.h [hunk 2]
```diff
@@ -336,6 +366,17 @@ inline std::string DLDataType2String(DLDataType t) {
   return os.str();
 }
 
+inline std::string VLADataType2String(DataType t) {
+  if (t.bits() == 0) return "";
+  std::ostringstream os;
+  os << t.operator DLDataType();
+  //  auto const str_to_parse =  os.str();
+  //  auto pos = str_to_parse.find("x");
+  //  auto stem= str_to_parse.substr(0, pos);
```
Comment (huajsj) [inline include/tvm/runtime/data_type.h:375]: remove

File: include/tvm/runtime/data_type.h [hunk 3]
```diff
@@ -69,14 +69,26 @@ class DataType {
...
+  //    data_.code = static_cast<uint8_t>(code);
+  //    data_.bits = static_cast<uint8_t>(bits);
+  //    is_scalable_ = true;
+  //    std::cout<<bits<<std::endl;
+  //    data_.lanes = uint16_t(128) / static_cast<uint16_t>(8); // minimal lanes
+  //
+  ////    if (code == kBFloat) {
+  ////      ICHECK_EQ(bits, 16);
+  ////    }
+  //  }
```
Comment (huajsj) [inline include/tvm/runtime/data_type.h:91]: remove

File: include/tvm/runtime/data_type.h [hunk 4]
```diff
@@ -107,6 +119,13 @@ class DataType {
   bool is_vector_bool() const { return is_vector() && bits() == 1; }
   /*! \return whether type is a Void type. */
   bool is_void() const { return code() == DataType::kHandle && bits() == 0 && lanes() == 0; }
+  bool is_scalable() const { return is_scalable_; }
+
+  DataType with_scalable_lanes() const {
+    int min_num_lanes = 128 / bits();
```
Comment (huajsj) [inline include/tvm/runtime/data_type.h:125]: macro? and does here only support vector size 128 or should be configurable between 128 - 2048?

File: include/tvm/runtime/data_type.h [hunk 5]
```diff
@@ -107,6 +119,13 @@ class DataType {
   bool is_vector_bool() const { return is_vector() && bits() == 1; }
   /*! \return whether type is a Void type. */
   bool is_void() const { return code() == DataType::kHandle && bits() == 0 && lanes() == 0; }
+  bool is_scalable() const { return is_scalable_; }
```
Comment (huajsj) [inline include/tvm/runtime/data_type.h:122]: doxygen comments.

File: include/tvm/runtime/data_type.h [hunk 6]
```diff
@@ -131,7 +150,7 @@ class DataType {
    */
   bool operator==(const DataType& other) const {
     return data_.code == other.data_.code && data_.bits == other.data_.bits &&
-           data_.lanes == other.data_.lanes;
+           data_.lanes == other.data_.lanes;  // && is_scalable_ == other.is_scalable_;
```
Comment (huajsj) [inline include/tvm/runtime/data_type.h:153]: seem like no diff with original code,  is the comment  for "is_scalable_ == other.is_scalable" intend logic?

File: include/tvm/runtime/data_type.h [hunk 7]
```diff
@@ -107,6 +119,13 @@ class DataType {
   bool is_vector_bool() const { return is_vector() && bits() == 1; }
   /*! \return whether type is a Void type. */
   bool is_void() const { return code() == DataType::kHandle && bits() == 0 && lanes() == 0; }
+  bool is_scalable() const { return is_scalable_; }
+
+  DataType with_scalable_lanes() const {
```
Comment (huajsj) [inline include/tvm/runtime/data_type.h:124]: how this function cooperate with is_scalable_? what happen if is_scalable_ is false but this function get called?

**Review by huajsj — COMMENTED** [review]
Thanks @mbaret.

**Comment by jroesch** [comment]
This PR appears to be out of date, please feel free to reopen it if this is not the case.

As part of the new year we are attempting to triage the project's open pull requests to ensure that code which
is ready for review and/or merging receives adequate attention.

Thanks again for your contribution, and feel free to reach out to discuss these changes.

**Pull request closed without merging**